Add looping and concurrency to `tpch`

cuongdo opened the pull request:
While we await the HTAP benchmarking goodness of CH-benCHmark, it's
useful to use TPC-H queries to load test DistSQL. This change adds the
ability to loop over the set of queries in multiple goroutines. This is
somewhat crude, because I don't see TPC-H being our long-term load
testing tool of choice.

Also converted various `fmt.Printf`s to `log.Printf`s, because
timestamps are useful with long-running tests.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/loadgen/43)
<!-- Reviewable:end -->

Commit:
Add looping and concurrency to `tpch`

While we await the HTAP benchmarking goodness of CH-benCHmark, it's
useful to use TPC-H queries to load test DistSQL. This change adds the
ability to loop over the set of queries in multiple goroutines. This is
somewhat crude, because I don't see TPC-H being our long-term load
testing tool of choice.

Also converted various `fmt.Printf`s to `log.Printf`s, because
timestamps are useful with long-running tests.

rjnn commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[tpch/main.go, line 219 at r1](https://reviewable.io:443/reviews/cockroachdb/loadgen/43#-Kh3ogCj_acCMSRTLJk9:-Kh3ogCj_acCMSRTLJkA:b6h55ko) ([raw file](https://github.com/cockroachdb/loadgen/blob/e78beed755da4566b710553a1e1dd621764ac247/tpch/main.go#L219)):*
> ```Go
> 		queries = append(queries, queryInt)
> 	}
> 	errChan := make(chan error, len(queries))
> ```

why `len(queries)`?

---

*[tpch/main.go, line 224 at r1](https://reviewable.io:443/reviews/cockroachdb/loadgen/43#-Kh3ow1w2opoQUADi2nc:-Kh3ow1w2opoQUADi2nd:b7f74bw) ([raw file](https://github.com/cockroachdb/loadgen/blob/e78beed755da4566b710553a1e1dd621764ac247/tpch/main.go#L224)):*
> ```Go
> 	}
> 	for i := uint(0); i < *concurrency; i++ {
> 		if err := <-errChan; err != nil {
> ```

If we're going to do continuous load, we probably want a flag for `max-fails` as well, and only do `log.Fatal` when we count up to that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/loadgen/43#-:-Kh3pJGJAaDCuFCb9olG:bnfp4nl)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions.

---

*[tpch/main.go, line 219 at r1](https://reviewable.io:443/reviews/cockroachdb/loadgen/43#-Kh3ogCj_acCMSRTLJk9:-Kh3w6sq9xJdQTS6tShP:b-aoy5b5) ([raw file](https://github.com/cockroachdb/loadgen/blob/e78beed755da4566b710553a1e1dd621764ac247/tpch/main.go#L219)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

why `len(queries)`?
</blockquote></details>

Force of habit. Removed the buffering, because it isn't needed.

---

*[tpch/main.go, line 224 at r1](https://reviewable.io:443/reviews/cockroachdb/loadgen/43#-Kh3ow1w2opoQUADi2nc:-Kh3wA5u4woMHsJB9AdJ:b1k2rup) ([raw file](https://github.com/cockroachdb/loadgen/blob/e78beed755da4566b710553a1e1dd621764ac247/tpch/main.go#L224)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

If we're going to do continuous load, we probably want a flag for `max-fails` as well, and only do `log.Fatal` when we count up to that.
</blockquote></details>

Added `max-errors` and did a bit of shuffling of code to make things work.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/loadgen/43)*
<!-- Sent from Reviewable.io -->

rjnn commented:
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[tpch/main.go, line 57 at r2](https://reviewable.io:443/reviews/cockroachdb/loadgen/43#-Kh46mcuOJTxVUe9qoDy:-Kh46mcuOJTxVUe9qoDz:b-xlh9rr) ([raw file](https://github.com/cockroachdb/loadgen/blob/a707f5617c640b7d5fae0a338cc1e924c28399fd/tpch/main.go#L57)):*
> ```Go
> var loops = flag.Uint("loops", 1, "Number of times to run the queries (0 = run forever).")
> var concurrency = flag.Uint("concurrency", 1, "Number of queries to execute concurrently.")
> var maxErrors = flag.Uint64("max-errors", 0, "Number of query errors allowed before aborting.")
> ```

The usual semantics is for the default to be `max-errors=1`, and  to use `<` when checking instead of `<=`, and `0` is used to indicate unlimited errors (just as in `var loops`).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/loadgen/43)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

*[tpch/main.go, line 57 at r2](https://reviewable.io:443/reviews/cockroachdb/loadgen/43#-Kh46mcuOJTxVUe9qoDy:-Kh4BUm6P9TwNo2PhsDD:b-896fix) ([raw file](https://github.com/cockroachdb/loadgen/blob/a707f5617c640b7d5fae0a338cc1e924c28399fd/tpch/main.go#L57)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

The usual semantics is for the default to be `max-errors=1`, and  to use `<` when checking instead of `<=`, and `0` is used to indicate unlimited errors (just as in `var loops`).
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/loadgen/43)*
<!-- Sent from Reviewable.io -->

rjnn commented:
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/loadgen/43)*
<!-- Sent from Reviewable.io -->
